File: src/utils/html-utils.ts
/**
 * HTML utility functions for the PDF-to-HTML pipeline.
 *
 * After switching rawText storage from plain text to structured HTML,
 * the KG pipeline (concept extraction, sentence splitting, entity
 * resolution, enrichment) and QA endpoints still need plain text.
 * These utilities provide safe HTML stripping and escaping.
 */

/**
 * Strip all HTML tags and return plain text.
 * - Converts <br> and </p> to newlines for readability
 * - Decodes common HTML entities
 * - Collapses excessive whitespace
 */
export function stripHtml(html: string): string {
  if (!html) return "";

  return html
    // Convert block-level closing tags to newlines
    .replace(/<\/(p|div|h[1-6]|li|tr|blockquote)>/gi, "\n")
    // Convert <br> to newlines
    .replace(/<br\s*\/?>/gi, "\n")
    // Convert <td>/<th> to tab-separated
    .replace(/<\/?(td|th)>/gi, "\t")
    // Remove all remaining tags
    .replace(/<[^>]+>/g, "")
    // Decode common HTML entities
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'")
    .replace(/&nbsp;/g, " ")
    .replace(/&#(\d+);/g, (_, code) => String.fromCharCode(parseInt(code, 10)))
    // Collapse 3+ newlines to 2
    .replace(/\n{3,}/g, "\n\n")
    // Trim each line
    .split("\n")
    .map((line) => line.trim())
    .join("\n")
    .trim();
}

/**
 * Escape HTML special characters for safe inline rendering.
 * Used when inserting PDF-extracted text into HTML structure.
 */
export function escapeHtml(text: string): string {
  if (!text) return "";
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

/**
 * Check if a string looks like HTML (contains tags).
 * Used to determine whether to strip before LLM processing.
 */
export function isHtml(text: string): boolean {
  if (!text) return false;
  return /<[a-z][\s\S]*?>/i.test(text);
}

/**
 * Truncate an HTML string to a maximum byte length WITHOUT leaving
 * open tags. Naive `substring(0, N)` cuts in the middle of `<figure>
 * <img src="...">` and the browser drops the broken element — this is
 * the root cause of missing images in OriginalTextPanel.
 *
 * Strategy:
 *  1. If the string is under the limit, return as-is.
 *  2. Otherwise, walk the string tracking open/close tags. Cut at the
 *     first position <= limit that is NOT inside a tag.
 *  3. Collect still-open tags and append their closing tags in reverse
 *     order so the result is well-formed HTML.
 *
 * The limit is on UTF-16 code unit length (String.length), which is
 * what the DB column size check uses anyway.
 */
export function safeTruncateHtml(html: string, maxLen: number): string {
  if (!html || html.length <= maxLen) return html || "";

  // Walk the string, tracking open tags. We only close void elements
  // list to skip them when emitting closers.
  const voidTags = new Set([
    "area", "base", "br", "col", "embed", "hr", "img", "input",
    "link", "meta", "param", "source", "track", "wbr",
  ]);

  const openStack: string[] = [];
  let i = 0;
  let cutPos = -1;

  while (i < html.length && i < maxLen) {
    const ch = html[i];

    if (ch === "<") {
      // Find the closing '>'
      const end = html.indexOf(">", i);
      if (end === -1) {
        // Malformed — cut here
        cutPos = i;
        break;
      }
      // Check if this is a closing tag </name>
      const tagContent = html.slice(i + 1, end);
      const isClosing = tagContent.startsWith("/");
      const tagNameMatch = tagContent.match(/^\/?\s*([a-zA-Z][a-zA-Z0-9]*)/);
      const tagName = tagNameMatch ? tagNameMatch[1].toLowerCase() : "";

      if (isClosing) {
        // Pop from stack until we find the matching tag
        const idx = openStack.lastIndexOf(tagName);
        if (idx !== -1) {
          openStack.splice(idx);
        }
      } else if (!voidTags.has(tagName) && !tagContent.endsWith("/")) {
        // Opening tag (non-void, non-self-closing)
        openStack.push(tagName);
      }

      // If this tag ends exactly at maxLen, cut after it
      if (end + 1 <= maxLen) {
        i = end + 1;
        cutPos = i;
      } else {
        // Tag straddles the limit — cut before this tag to keep it whole
        cutPos = i;
        break;
      }
    } else {
      i++;
      if (i <= maxLen) cutPos = i;
    }
  }

  if (cutPos === -1) cutPos = maxLen;

  let result = html.slice(0, cutPos);

  // Append closers for any still-open tags, in reverse order
  for (let j = openStack.length - 1; j >= 0; j--) {
    result += `</${openStack[j]}>`;
  }

  return result;
}

File: python-parser/app/utils/html_utils.py
"""
HTML 工具函数，对应 src/utils/html-utils.ts。

KG 流水线前需要 stripHtml 把结构化 HTML（PDF/DOCX 产物）转成纯文本送给 LLM；
escapeHtml 用于解析器把提取出的文本插入 HTML 结构时转义。
"""
import re
from typing import Set

# void 标签集合（safe_truncate_html 用）—— 与 TS 版本一致
_VOID_TAGS: Set[str] = {
    "area", "base", "br", "col", "embed", "hr", "img", "input",
    "link", "meta", "param", "source", "track", "wbr",
}


def strip_html(html: str) -> str:
    """去除所有 HTML 标签，返回纯文本。

    - block 级闭合标签转换为换行
    - <br> 转换为换行
    - <td>/<th> 转换为 tab
    - 解码常见 HTML 实体
    - 折叠 3+ 换行为 2 个
    """
    if not html:
        return ""

    s = html
    # block 级闭合标签 → 换行
    s = re.sub(r'</(p|div|h[1-6]|li|tr|blockquote)>', '\n', s, flags=re.IGNORECASE)
    # <br> → 换行
    s = re.sub(r'<br\s*/?>', '\n', s, flags=re.IGNORECASE)
    # <td>/<th> → tab
    s = re.sub(r'</?(td|th)>', '\t', s, flags=re.IGNORECASE)
    # 移除所有剩余标签
    s = re.sub(r'<[^>]+>', '', s)
    # 解码常见 HTML 实体（顺序与 TS 版本一致，&amp; 必须最先）
    s = s.replace('&amp;', '&')
    s = s.replace('&lt;', '<')
    s = s.replace('&gt;', '>')
    s = s.replace('&quot;', '"')
    s = s.replace('&#39;', "'")
    s = s.replace('&apos;', "'")
    s = s.replace('&nbsp;', ' ')
    # 数字实体 &#NNN;
    s = re.sub(r'&#(\d+);', lambda m: chr(int(m.group(1))), s)
    # 折叠 3+ 换行为 2 个
    s = re.sub(r'\n{3,}', '\n\n', s)
    # 每行去首尾空白
    s = '\n'.join(line.strip() for line in s.split('\n'))
    return s.strip()


def escape_html(text: str) -> str:
    """转义 HTML 特殊字符，用于把提取文本安全插入 HTML 结构。"""
    if not text:
        return ""
    s = text
    s = s.replace('&', '&amp;')
    s = s.replace('<', '&lt;')
    s = s.replace('>', '&gt;')
    s = s.replace('"', '&quot;')
    s = s.replace("'", '&#39;')
    return s


def is_html(text: str) -> bool:
    """检测字符串是否像 HTML（包含标签）。用于决定是否 strip 后再送 LLM。"""
    if not text:
        return False
    return bool(re.search(r'<[a-z][\s\S]*?>', text, re.IGNORECASE))


def safe_truncate_html(html: str, max_len: int) -> str:
    """安全截断 HTML 到最大长度，不留下未闭合标签。

    朴素的 substring(0, N) 会切断 <figure><img src="..."> 中间，
    导致浏览器丢弃破损元素——这是 OriginalTextPanel 图片丢失的根因。

    策略（与 TS 版本逐行对齐）：
      1. 长度未超限直接返回
      2. 遍历字符串追踪开/闭标签栈，在不超过 limit 且不在标签内部的位置截断
      3. 收集仍打开的标签，按逆序追加闭合标签
    """
    if not html or len(html) <= max_len:
        return html or ""

    open_stack: list[str] = []
    i = 0
    cut_pos = -1

    while i < len(html) and i < max_len:
        ch = html[i]

        if ch == '<':
            end = html.find('>', i)
            if end == -1:
                # 畸形标签——在此截断
                cut_pos = i
                break
            tag_content = html[i + 1:end]
            is_closing = tag_content.startswith('/')
            tag_name_match = re.match(r'^/?\s*([a-zA-Z][a-zA-Z0-9]*)', tag_content)
            tag_name = tag_name_match.group(1).lower() if tag_name_match else ""

            if is_closing:
                # 从栈中弹出直到匹配的标签
                try:
                    idx = len(open_stack) - 1 - open_stack[::-1].index(tag_name)
                    open_stack = open_stack[:idx]
                except ValueError:
                    pass  # 栈中没有匹配项，忽略
            elif tag_name not in _VOID_TAGS and not tag_content.endswith('/'):
                # 开标签（非 void、非自闭合）
                open_stack.append(tag_name)

            # 如果标签整体在 max_len 内，跳过它
            if end + 1 <= max_len:
                i = end + 1
                cut_pos = i
            else:
                # 标签跨越 limit——在此标签前截断以保持完整
                cut_pos = i
                break
        else:
            i += 1
            if i <= max_len:
                cut_pos = i

    if cut_pos == -1:
        cut_pos = max_len

    result = html[:cut_pos]
    # 按逆序追加未闭合标签
    for j in range(len(open_stack) - 1, -1, -1):
        result += f'</{open_stack[j]}>'

    return result
